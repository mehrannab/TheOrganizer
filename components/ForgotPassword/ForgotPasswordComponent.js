import { StyleSheet, View, ToastAndroid, Dimensions } from 'react-native'
import React, {useState} from 'react'
import {Button, Headline, Text, TextInput} from 'react-native-paper';
import { getAuth, sendPasswordResetEmail } from "firebase/auth";
import { useNavigation } from '@react-navigation/native';
import { ScaledSheet } from 'react-native-size-matters';

const {height, width} = Dimensions.get('window')

export default function ForgotPasswordComponent() {
    const [email, setEmail] = useState('');
    const auth = getAuth();
    const navigation = useNavigation();


    const forgotPassword = () => {
    sendPasswordResetEmail(auth, email)
        .then(() => {
            ToastAndroid.show('Reset password mail sent successfully!', ToastAndroid.LONG);
        })
        .catch(() => {
            ToastAndroid.show('Error: Please check your entered e-mail!', ToastAndroid.LONG)
        });
    }


  return (
    <View style={styles.container}>
        <View>
            <Button style={styles.buttonBack} uppercase={false} onPress={() => {navigation.navigate("Login")}}>
                <Text style={styles.buttonTextGoBack}>Go back</Text>
            </Button>
        </View>
        <View>
            <Headline style={styles.headline}>Forgot your password? No problem!</Headline>
        </View>
        <View>
            <Text style={styles.inputText}>Email:</Text>
            <TextInput style={styles.input} placeholder="Enter your email" value={email} onChangeText={text => setEmail(text)}></TextInput>
        </View>
        <View>
            <Button testID='ResetPassword.Button' style={styles.buttonResetPassword} mode="contained" uppercase={false} onPress={forgotPassword}>
                <Text style={styles.buttonTextResetPassword}>Reset password</Text>
            </Button>
        </View>

    </View>
  )
}


const styles = ScaledSheet.create({
    container: {
        flex: 1,
        alignItems: 'center',
        justifyContent: 'center',
        marginBottom: '200@s'
    },
    headline: {
        fontSize: '18@s',
        marginTop: '60@s',
        padding: '20@s',
    },
    inputText: {
        paddingLeft: '14@s',
        paddingTop: '6@s',
        fontSize: '14@s'
    },
    input: {
        padding: '10@s',
        margin: '15@s',
        width: width * 0.8,
        height: height * 0.03,
        fontSize: '14@s'
    },
    buttonResetPassword: {
        width: width * 0.5,
    },
    buttonBack: {
        paddingRight: '260@s',
        width: width * 1.1,
    },
    buttonTextResetPassword: {
        color: 'white',
        fontSize: '14@s'
    },
    buttonTextGoBack: {
        fontSize: '14@s'
    }
})