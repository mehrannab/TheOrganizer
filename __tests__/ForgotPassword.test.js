import React from 'react';
import renderer, { act } from 'react-test-renderer'
import ForgotPasswordComponent from '../components/ForgotPassword/ForgotPasswordComponent';
import { NavigationContainer } from '@react-navigation/native';
import { fireEvent, render, waitFor } from '@testing-library/react-native';
import { Button, TextInput } from 'react-native-paper';
jest.useFakeTimers()
jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');

const mockGetAuth = jest.fn()

const mockSendPassword = jest.fn(() => {
    return Promise.resolve(true)
})

jest.mock('firebase/auth', () => {
    return {
        getAuth: () => mockGetAuth,
        sendPasswordResetEmail: () => mockSendPassword
    }
})



// Snapshot testing
test('renders forgotPasswordComponent correctly 1', () => {
    const tree = renderer.create(
    <NavigationContainer>
    <ForgotPasswordComponent/>
    </NavigationContainer>)
    .toJSON();
    expect(tree).toMatchSnapshot();
});



//Component rendering testing - Specific widgets
describe('forgotPasswordComponent', () => {
    test("renders default element", () => {
        const {getAllByText, getByPlaceholderText} = render(
        <NavigationContainer>
        <ForgotPasswordComponent/>
        </NavigationContainer>
        );
        
        expect(getAllByText('Forgot your password? No problem!').length).toBe(1);
        expect(getAllByText('Email:').length).toBe(1);
        getByPlaceholderText("Enter your email");
    });
});



//Testing buttons and inputfield
describe('Buttons and input on press working', () => {
    it('Go back onpress is called', () => {
        const onPress = jest.fn();
        const {getByText} = render(
            <Button onPress={onPress}>Go Back</Button>,
        );
        fireEvent.press(getByText('Go Back'));
        expect(onPress).toBeCalledTimes(1);
    });

    it('Reset onpress is called', () => {
        const onPress = jest.fn();
        const {getByText} = render(
            <Button onPress={onPress}>Reset password</Button>,
        )
        fireEvent.press(getByText('Reset password'));
        expect(onPress).toBeCalledTimes(1);
    })

    it('Email onChange is called', () => {
        const onChangeTextMock = jest.fn();
        const {getByPlaceholderText} = render(
            <TextInput placeholder="Enter your email" onChangeText={onChangeTextMock}/>
        );
        fireEvent.changeText(getByPlaceholderText("Enter your email"), "test@gmail.com");
        expect(onChangeTextMock).toBeCalled();
    })
});



//Testing that navigation is called when Go-Back button is pressed
describe("Navigation working", () => {
    it("calls useNavigation and navigates", () => {
        const mockedNavigate = jest.fn();
        jest.mock('@react-navigation/native', () => (
        { useNavigation: () => ({ navigate: mockedNavigate }) }));        

        const onPress = jest.fn(mockedNavigate);
        const {getByText} = render(
            <Button onPress={onPress}>Go Back</Button>,
        );

        fireEvent.press(getByText('Go Back'));
        expect(mockedNavigate).toBeCalledTimes(1);
    })
})


