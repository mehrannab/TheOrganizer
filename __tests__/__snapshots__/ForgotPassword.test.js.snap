// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders forgotPasswordComponent correctly 1 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
      "marginBottom": 428.57142857142856,
    }
  }
>
  <View>
    <View
      style={
        Object {
          "backgroundColor": "transparent",
          "borderColor": "transparent",
          "borderRadius": 4,
          "borderStyle": "solid",
          "borderWidth": 0,
          "elevation": 0,
          "minWidth": 64,
          "paddingRight": 557.1428571428571,
          "width": 825.0000000000001,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "disabled": undefined,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 4,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              },
              undefined,
            ]
          }
        >
          <Text
            numberOfLines={1}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "letterSpacing": 1,
                    "marginHorizontal": 16,
                    "marginVertical": 9,
                    "textAlign": "center",
                  },
                  undefined,
                  false,
                  Object {
                    "color": "#6200ee",
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  Object {
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  undefined,
                ],
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Object {
                    "fontSize": 30,
                  },
                ]
              }
            >
              Go back
            </Text>
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View>
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "System",
            "fontWeight": "400",
          },
          Object {
            "textAlign": "left",
          },
          Array [
            Object {
              "textAlign": "left",
            },
            Object {
              "color": "rgba(0, 0, 0, 0.87)",
              "fontFamily": "System",
              "fontWeight": "400",
              "writingDirection": "ltr",
            },
            Array [
              Object {
                "fontSize": 24,
                "letterSpacing": 0,
                "lineHeight": 32,
                "marginVertical": 2,
              },
              Object {
                "fontSize": 38.57142857142857,
                "marginTop": 128.57142857142856,
                "padding": 42.857142857142854,
              },
            ],
          ],
        ]
      }
    >
      Forgot your password? No problem!
    </Text>
  </View>
  <View>
    <Text
      style={
        Array [
          Object {
            "color": "#000000",
            "fontFamily": "System",
            "fontWeight": "400",
          },
          Object {
            "textAlign": "left",
          },
          Object {
            "fontSize": 30,
            "paddingLeft": 30,
            "paddingTop": 12.857142857142858,
          },
        ]
      }
    >
      Email:
    </Text>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgb(231, 231, 231)",
            "borderTopLeftRadius": 4,
            "borderTopRightRadius": 4,
          },
          Object {
            "margin": 32.14285714285714,
            "padding": 21.428571428571427,
            "width": 600,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.26)",
            "bottom": 0,
            "height": 2,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "transform": Array [
              Object {
                "scaleY": 0.5,
              },
            ],
            "zIndex": 1,
          }
        }
      />
      <View
        style={
          Array [
            Object {
              "paddingBottom": 0,
              "paddingTop": 0,
            },
            Object {
              "minHeight": 40.019999999999996,
            },
          ]
        }
      >
        <TextInput
          allowFontScaling={true}
          editable={true}
          maxFontSizeMultiplier={1.5}
          multiline={false}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="Enter your email"
          placeholderTextColor="rgba(0, 0, 0, 0.54)"
          rejectResponderTermination={true}
          selectionColor="#6200ee"
          style={
            Array [
              Object {
                "flexGrow": 1,
                "margin": 0,
              },
              Object {
                "paddingLeft": 12,
                "paddingRight": 12,
              },
              Object {
                "height": 40.019999999999996,
              },
              Object {
                "paddingBottom": 0,
                "paddingTop": 0,
              },
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontSize": 30,
                "fontWeight": undefined,
                "textAlign": "left",
                "textAlignVertical": "center",
              },
              false,
              Array [
                Object {},
              ],
            ]
          }
          testID="text-input-flat"
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
  </View>
  <View>
    <View
      style={
        Object {
          "backgroundColor": "#6200ee",
          "borderColor": "transparent",
          "borderRadius": 4,
          "borderStyle": "solid",
          "borderWidth": 0,
          "elevation": 2,
          "minWidth": 64,
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 0.75,
            "width": 0,
          },
          "shadowOpacity": 0.24,
          "shadowRadius": 1.5,
          "width": 375,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "disabled": undefined,
          }
        }
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 4,
            },
          ]
        }
        testID="ResetPassword.Button"
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              },
              undefined,
            ]
          }
        >
          <Text
            numberOfLines={1}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "letterSpacing": 1,
                    "marginHorizontal": 16,
                    "marginVertical": 9,
                    "textAlign": "center",
                  },
                  undefined,
                  false,
                  Object {
                    "color": "#ffffff",
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  Object {
                    "fontFamily": "System",
                    "fontWeight": "500",
                  },
                  undefined,
                ],
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Object {
                    "color": "white",
                    "fontSize": 30,
                  },
                ]
              }
            >
              Reset password
            </Text>
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
